[jit] Support pack_padded_sequence and pad_packed_sequence

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23266 [jit] Support nn.GRU in script
* **#23249 [jit] Support pack_padded_sequence and pad_packed_sequence**
* #23248 [jit] support torch._C._get_tracing_state in script
* #23247 [jit] support torch.as_tensor in script

Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)

wanchaol committed:
[jit] Support pack_padded_sequence and pad_packed_sequence

wanchaol committed:
refactor on "[jit] Support pack_padded_sequence and pad_packed_sequence"

[jit] Support pack_padded_sequence and pad_packed_sequence

gh-metadata: pytorch pytorch 23249 gh/wanchaol/34/head

wanchaol committed:
Update on "[jit] Support pack_padded_sequence and pad_packed_sequence"

wanchaol committed:
Update on "[jit] Support pack_padded_sequence and pad_packed_sequence"


Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)

eellison commented on test/test_jit.py:
```diff
@@ -7736,6 +7736,27 @@ def forward(self, x, seq_lens):
+            if seq_lens[b] < T:
+                x[seq_lens[b]:, b, :] = 0
+
+        eager_seq = pack_padded_pad_packed_script(x, seq_lens)
```
reason not to do checkScript here instead ?

wanchaol replied:
`CheckScript` is also checking the string frontend and the string frontend could not handle the comments starting with `r`..

eellison commented on test/test_jit.py:
```diff
@@ -7736,6 +7736,27 @@ def forward(self, x, seq_lens):
+
+        def pack_padded_pad_packed_script(x, seq_lens):
+            x = pack_padded_sequence(x, seq_lens)
+            x, _ = pad_packed_sequence(x)
```
check both outputs here ?

eellison reviewed (approved):
LGTM. maybe add a couple more test inputs

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -254,7 +261,7 @@ def pack_padded_sequence(input, lengths, batch_first=False, enforce_sorted=True)
                       'values, and it will treat them as constants, likely rendering '
                       'the trace incorrect for any other combination of lengths.',
-                      category=torch.jit.TracerWarning, stacklevel=2)
+                      stacklevel=2)
```
Can't we make JIT ignore the category? It's not a good idea to remove it

eellison replied:
category is hard to ignore atm because we don't have any way of typing `torch.jit.TracerWarning`

wanchaol replied:
yeah I tried that and it's kind of hard now, plus other callsites does not have this category either way.. not sure if someone deleted them before.

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -310,12 +318,12 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0, total_le
         batch_dim = 0 if batch_first else 1
-        return padded_output.index_select(batch_dim, sequence.unsorted_indices), \
-            lengths[sequence.unsorted_indices]
+        return padded_output.index_select(batch_dim, unsorted_indices), lengths[unsorted_indices]
```
Why did you have to rewrite this code? Because of optional unwrapping?

wanchaol replied:
Yes, our type refinement now is limited to TK_VAR only, we will need to expand it to a general case, see https://github.com/pytorch/pytorch/issues/23049

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -310,12 +318,12 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0, total_le
                              .format(total_length, max_seq_length))
         max_seq_length = total_length
-    padded_output, lengths = torch._C._VariableFunctions._pad_packed_sequence(
+    padded_output, lengths = _VF._pad_packed_sequence(
```
Is this because the JIT can't handle `torch._C._VariableFunctions`?

wanchaol replied:
Yes, we are using a wrapper for `_VariableFunctions` (which is _VF) for it to be accessible in JIT.

wanchaol committed:
Update on "[jit] Support pack_padded_sequence and pad_packed_sequence"


Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)

wanchaol committed:
Update on "[jit] Support pack_padded_sequence and pad_packed_sequence"


Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)

wanchaol committed:
refactor on "[jit] Support pack_padded_sequence and pad_packed_sequence"


Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)

wanchaol committed:
Update on "[jit] Support pack_padded_sequence and pad_packed_sequence"


Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)

wanchaol committed:
rebase with master on "[jit] Support pack_padded_sequence and pad_packed_sequence"


Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)
